components/sup/src/manager/mod.rs

```diff
@@ -934,138 +1096,213 @@ impl Manager {
-            mem::swap(services.deref_mut(), &mut svcs);
-        }
+                .expect("Watcher lock poisoned")
+                .remove(&service);
+            updater
+                .lock()
+                .expect("Updater lock poisoned")
+                .remove(&service);
+            Ok(())
+        });
+        Self::wrap_async_service_operation(ident, busy_services, reconcile_services, stop_it)
+    }
+
+    /// Wrap a future that starts, stops, or restarts a service with
+    /// logic that marks that service as "busy" for the duration of
+    /// the process.
+    ///
+    /// This allows us to postpone taking additional action on
+    /// services until they're done with what they're doing.  (For
+    /// example, consider stopping a service that takes 10 seconds to
+    /// shut down (including post-stop hook execution), but then
+    /// executing `hab svc start SERVICE` 2 seconds into that 10
+    /// seconds.)
+    ///
+    /// As more service operations (e.g., hooks) become asynchronous,
+    /// we'll need to wrap those operations in this logic to ensure
+    /// consistent operation.
+    fn wrap_async_service_operation<F>(
+        ident: PackageIdent,
+        busy_services: Arc<Mutex<HashSet<PackageIdent>>>,
+        reconcile_services: Arc<AtomicBool>,
+        fut: F,
+    ) -> impl Future<Item = (), Error = ()>
+    where
+        F: IntoFuture<Item = (), Error = ()>,
+    {
+        // TODO (CM): can't wait for the Pinning API :(
+        let busy_services_2 = busy_services.clone();
+        let ident_2 = ident.clone();
+
+        future::lazy(move || {
+            trace!(
+                "Flagging '{:?}' as busy, pending an asynchronous operation",
+                ident
+            );
+            busy_services
+                .lock()
+                .expect("busy_services lock is poisoned")
+                .insert(ident);
+            Ok(())
+        })
+        .and_then(|_| fut)
+        .and_then(move |_| {
+            trace!(
+                "Removing 'busy' flag for '{:?}'; asynchronous operation over",
+                ident_2
+            );
+            busy_services_2
+                .lock()
+                .expect("busy_services lock is poisoned")
+                .remove(&ident_2);
+            reconcile_services.store(true, Ordering::Relaxed);
+            Ok(())
+        })
+    }
+
+    /// Indicates if we need to examine our on-disk specfiles in order
+    /// to reconcile them with whatever we're currently running.
+    ///
+    /// Takes into account filesystem events in the specs directory,
+    /// as well as whether or not we need to reexamine specs after
+    /// finishing some asynchronous operation on a service.
+    fn need_to_reconcile_spec_files(&mut self) -> bool {
+        self.spec_watcher.has_events()
+            || self
+                .reconcile_services
+                .compare_and_swap(true, false, Ordering::Relaxed)
```

Comment:
I'm not quite following. `reconcile_services` is set to `true` by futures as they complete, and only set to `false` if we're ticking through the main loop and see that it's currently got a value of `true`, which is our signal that "something updated since the last time we went through".

Actually, the only way that `maybe_spawn_service_futures` can fail is if there are spec files that either can't be read or contain non-UTF-8 bytes... see [this][1] and ultimately [this][2]. Given that we simply log when spec files can't be properly parsed, or contain an incorrect ident, I don't see how it's useful to bubble up an error in the other cases, particularly when that is the reason that a lot of the code in the manager loop needs to deal with `Result` at all.

[1]: https://github.com/habitat-sh/habitat/blob/6e44bc89d140862cbc75c92f12fb25c8b1d85eee/components/sup/src/manager/spec_dir.rs#L87-L89
[2]: https://github.com/habitat-sh/habitat/blob/6e44bc89d140862cbc75c92f12fb25c8b1d85eee/components/sup/src/manager/service/spec.rs#L188-L198

Reply:
Since we have this caller
```rust
            if self.need_to_reconcile_spec_files() {
                self.maybe_spawn_service_futures(&mut runtime)?;
            }
```
It seems a bit unexpected that `need_to_reconcile_spec_files` is a self-mutating function. Even if the likelihood of failures is remote, this seems more natural to me:
```rust
            if self.need_to_reconcile_spec_files() {
                self.maybe_spawn_service_futures(&mut runtime)?;
                self.reconcile_services
                    .compare_and_swap(true, false, Ordering::Relaxed);
            }
```
and change`self.need_to_reconcile_spec_files` thusly:
```rust
    fn need_to_reconcile_spec_files(&self) -> bool {
        self.spec_watcher.has_events() || self.reconcile_services.load(Ordering::Relaxed)
    }
```
Does that make my question clearer?

Reply:
I think that if we split the checking and modification, we might open ourselves up to a race condition, where some service comes in to set it to true before we do the compare and swap, which could mean that we don't realize that we need to restart something for a considerably longer time than we should.

Reply:
Now *I'm* not following.

We do the `if self.need_to_reconcile_spec_files()` check every time through the main loop, right? My concern was that if we do the `compare_and_swap` first, we're saying we've done the work of reconciliation before we've actually completed it (or even committed to doing it in the future, necessarily). Since the next time through the loop, there's no indication that we didn't succeed in the reconciliation work (or scheduling that work with futures), we may just not do it until some other thing happens that triggers reconciliation.

Can you describe in more detail the downside to this approach?
```rust
            if self.need_to_reconcile_spec_files() { // non-mutating check
                self.maybe_spawn_service_futures(&mut runtime)?;
                self.reconcile_services
                    .compare_and_swap(true, false, Ordering::Relaxed);
            }
```
? I think I need to be walked through an example of how a race could lead to a bad outcome here.

Reply:
Sure thing.

If our flag is true, and we do a non-mutating check, then we get into the body of our `if`; all good. However, as we're constructing our futures to spawn, another service shutdown future registers that it has finished by setting the bool to `true`, but it's not taken into account in the futures that we're spawning. After we spawn the futures, we then set the flag to false, but we've effectively lost knowledge of that one service having finished shutting down. The next time we go through the loop, we'll see the bool as `false`, meaning that we're not going to reconcile that one service until either a) something changes a file in the spec directory, or b) another service finishes shutting down and flips the bool back to `true` as a result. Depending on the level of activity, that could be a very long time, which would be bad if that service was shutting down to restart.

The more I think about this, the more I really want to modify the spec generation to not throw an error (see https://github.com/habitat-sh/habitat/pull/6107#discussion_r256133998)... it seems like the possibility of failure in `maybe_spawn_service_futures` is the thing that you're keying off of.